### PR TITLE
Add WandB user guide

### PR DESCRIPTION
**Pull request by satyaog**
## Summary
Adds a new user guide for tracking experiments with Weights & Biases on the Mila cluster, covering SSO sign-in, CLI authentication, run initialization, metric logging, run organization, and I/O bottleneck diagnosis. Also updates the WandB example script with step timing, run grouping, and code snippet markers.

## Changes
- New `docs/userguides/wandb.md` guide covering SSO sign-in, CLI auth, `wandb.init()`, `wandb.log()`, run organization with `group=`/`tags=`, and bottleneck diagnosis using `perf/data_load_s` and `perf/compute_s` metrics.
- Updated `docs/examples/good_practices/wandb_setup/main.py` with step timing metrics, `group=`/`tags=` in `wandb.init()`, explicit `wandb.finish()`, validation logging, and `--8<--` code snippet markers for embedding in the guide.
- Added WandB guide entry to `docs/README.md` table of contents and `docs/userguides/index.md` table.
- Added "Track Experiments with WandB" next-step card to `docs/getting_started/train_first_model.md`.
- Minor copy improvements across Getting Started pages (description text and card labels).

### CODE REVIEW

File: docs/userguides/wandb.md [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+
+Most WandB Python API calls and sweep commands require a valid API key stored in
+the environment.
+
+### Log in interactively
+
+Use this on a login node or an interactive allocation for exploratory work:
+
+```bash
+uvx "wandb@latest" login
```
Comment (lebrice) [inline docs/userguides/wandb.md]:
```suggestion
uv tool install wandb
wandb login
```

File: docs/userguides/wandb.md [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+    same cluster do not need to re-run this command unless the key is rotated.
+
+### Offline mode
+
+On clusters without outbound internet access on compute nodes, run in offline
+mode and sync after the job completes:
+
+```bash
+export WANDB_MODE=offline
+srun uv run python main.py
```
Comment (lebrice) [inline docs/userguides/wandb.md]:
```suggestion
srun uv run --offline python main.py
```

File: docs/userguides/wandb.md [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+```
+
+After the job finishes, sync the run from the login node:
+
+```bash
+uvx "wandb@latest" sync --sync-all
```
Comment (lebrice) [inline docs/userguides/wandb.md]:
```suggestion
wandb sync --sync-all
```

File: docs/userguides/wandb.md [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+```python
+wandb.init(
+    project="wandb-example",
+    name=os.environ.get("SLURM_JOB_ID"),
+    group=os.environ.get("SLURM_ARRAY_JOB_ID", None),   # (1)!
+    resume="allow",
+    tags=["example", "resnet18"],                       # (2)!
+    config=vars(args),
+)
```
Comment (lebrice) [inline docs/userguides/wandb.md:229]: The ID needs to be the same for resuming to work correctly. If there are multiple tasks, then this would might cause an issue, but I don't think we need to worry about that here.
```suggestion
wandb.init(
    project="wandb-example",
    name=os.environ.get("SLURM_JOB_ID"),
    id=os.environ.get("SLURM_JOB_ID"),
    group=os.environ.get("SLURM_ARRAY_JOB_ID"),   # (1)!
    resume="allow",
    tags=["example", "resnet18"],                       # (2)!
    config=vars(args),
)
```

File: docs/userguides/wandb.md [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+```
+{ .annotate }
+
+1. Clusters multi-seed under a single expandable row. Passing
+   `SLURM_ARRAY_JOB_ID` groups every task in a job array together automatically.
```
Comment (lebrice) [inline docs/userguides/wandb.md]:
```suggestion
1. Using `SLURM_ARRAY_JOB_ID` automatically as the group clusters all jobs into a single expandable row.
```

File: docs/userguides/wandb.md [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+### Initialize a run
+
+```python
+import os
+import wandb
+
+wandb.init(
+    project="wandb-example",                # (1)!
+    name=os.environ.get("SLURM_JOB_ID"),    # (2)!
+    resume="allow",                         # (3)!
```
Comment (lebrice) [inline docs/userguides/wandb.md]:
```suggestion
    name=os.environ.get("SLURM_JOB_ID"),    # (2)!
    id=os.environ.get("SLURM_JOB_ID"),
    resume="allow",                         # (3)!
```

File: docs/userguides/wandb.md [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+{ .annotate }
+
+1. Clusters multi-seed under a single expandable row. Passing
+   `SLURM_ARRAY_JOB_ID` groups every task in a job array together automatically.
+2. Labels runs for filtering in the Runs table.
+
+## Diagnose training bottlenecks
+
+WandB records GPU utilization, CPU usage, and memory under the **System** tab of
+every run automatically — no extra code required. These metrics are the first
```
Comment (lebrice) [inline docs/userguides/wandb.md]: Les caractères `—` sont des signes de LLMs, personellement je les enlèverais.
```suggestion
every run automatically, no extra code is required. These metrics are the first
```

File: docs/userguides/wandb.md [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+    )
+```
+
+When `perf/data_load_s` is consistently larger than `perf/compute_s`, the job is
+I/O-bound.
+
+!!! tip
+    Common fixes for an I/O bottleneck: increase `num_workers` in the
+    `DataLoader`, enable `pin_memory=True`, or copy the dataset to
+    `$SLURM_TMPDIR` before the job starts.
```
Comment (lebrice) [inline docs/userguides/wandb.md]: Hmm. C'est pas une mauvaise idée, mais il faudrait aussi ajouter des `torch.cuda.synchronize()` pour faire ça comme il faut, et juste comme ça on risque d'induire un peu les gens en erreur. Je pense que de le faire bien compliquerait beaucoup l'exemple, et ça aurait plus à voir avec le profiling qu'avec wandb.

Je crois qu'il vaudrait mieux s'en tenir à l'identification du problème par le GPU util, et faire un guide plus approfondi sur le profiling eventuellement. Qu'en penses-tu?

File: docs/examples/good_practices/wandb_setup/main.py [hunk 9]
```diff
@@ -80,14 +81,18 @@ def main():
     # This specific example here does not do that.
 
     # Setup Wandb
+    # --8<-- [start:wandb-init]
     wandb.init(
         # Set the project where this run will be logged
         project="wandb-example",
         name=os.environ.get("SLURM_JOB_ID"),
+        group=os.environ.get("SLURM_ARRAY_JOB_ID", None),
         resume="allow",  # See https://docs.wandb.ai/guides/runs/resuming
```
Comment (lebrice) [inline docs/examples/good_practices/wandb_setup/main.py:90]:
```suggestion
        name=os.environ.get("SLURM_JOB_ID"),
        id=os.environ.get("SLURM_JOB_ID"),
        group=os.environ.get("SLURM_ARRAY_JOB_ID"),
        resume="allow",  # See https://docs.wandb.ai/guides/runs/resuming
```

File: docs/examples/good_practices/wandb_setup/main.py [hunk 10]
```diff
@@ -137,7 +142,16 @@ def main():
...
+        # captures any overhead before the loop begins.
+        t_step_end = time.perf_counter()
         for batch in train_dataloader:
+            # The dataloader has already fetched the batch by the time we reach
+            # here, so the elapsed time since t_step_end is the data-load wait.
+            t_data_load_end = time.perf_counter()
+            data_load_s = t_data_load_end - t_step_end
+
             # Move the batch to the GPU before we pass it to the model
             batch = tuple(item.to(device) for item in batch)
```
Comment (lebrice) [inline docs/examples/good_practices/wandb_setup/main.py]: The time taken to move the data to the GPU should also be included in the data load time IMO. To do this properly would also require a torch cuda synchronize, which would complicate the example.

Comment (lebrice) [reply docs/examples/good_practices/wandb_setup/main.py]: Also, if we do a profiling example eventually, I'd use pinned memory in the dataloader, use `item.to(device, non_blocking=True)` to load things asynchronously to the GPU, and use torch cuda events to segment out the different parts.

**Review by lebrice — APPROVED** [review]
Nice!

J'ai ajouté des suggestions, nottamment d'utiliser le `id=os.environ.get("SLURM_JOB_ID")` pour que le resume fonctionne correctement.

Je crois que le côté profiling du guide est un peu out-of-scope, et on pourrait le faire plus en détail (et en utilisant les trucs comme torch cuda synchronize, cuda events, async loading, etc) dans un guide dédié à ça. 
Par contre, présentement je crois que tout est mesuré comme il faut simplement parce qu'il y a autant d'appels à des trucs comme `loss.item()` dans le training script, (ce qui est ma faute). Du moment qu'on optimisera le code des exemples pour ne plus faire ça, il faudra ajouter des cuda sync, et le code qui mesure ces métriques de perf devra être adaptées un peu.

On peut le laisser pour l'instant. 

Aussi, tu verras que le CI fail parce que le output vers la fin du job (quand wandb upload les résultats) est inclu dans le regression check. On pourrait juste chopper la fin du output si tu veux.

File: docs/userguides/wandb.md [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+
+```python
+import os
+import wandb
+
+wandb.init(
+    project="wandb-example",                # (1)!
+    name=os.environ.get("SLURM_JOB_ID"),    # (2)!
+    resume="allow",                         # (3)!
+    config=vars(args),                      # (4)!
```
Comment (lebrice) [inline docs/userguides/wandb.md]:
```suggestion
    config=vars(args) | {f"env/{k}": v for k, v in os.environ.items() if k.startswith("SLURM"},    # (4)!
```

File: docs/userguides/wandb.md [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+```
+{ .annotate }
+
+1. Groups runs in the WandB UI. Use one project per research question.
+2. Links the run to its log file (`slurm-<JOBID>.out`) for easy
+   cross-referencing.
+3. Creates a new run if the ID does not exist, or resumes it if it does. Useful
+   when combined with checkpointing to recover from preemption.
+4. Pass the full `argparse` namespace — every key becomes a searchable,
+   filterable column in the WandB Runs table under **Config**.
```
Comment (lebrice) [inline docs/userguides/wandb.md]:
```suggestion
4. Pass the full `argparse` namespace and SLURM environment variables for easier debugging.
    Every key becomes a searchable, filterable column in the WandB Runs table under **Config**.
```

File: docs/examples/good_practices/wandb_setup/main.py (hunk 10 above)
Comment (satyaog) [reply docs/examples/good_practices/wandb_setup/main.py]: Reverted. I'm not sure about a torch cuda synchronize, it seams like a thing to avoid to do in general no? In the current example, I, as you do, think there are implicit cuda sync since the f-strings in the `logger.debug` are evaluated in for each batch. In a profiling example we should indeed probably put in there all the best practice to avoid starving the GPU

File: docs/userguides/wandb.md (hunk 8 above)
Comment (satyaog) [reply docs/userguides/wandb.md]: Reverted, mais je pense, comme toi aussi, qu'il y a déjà des cuda sync implicite par les f-strings dans les `logger.debug`.

**Comment by lebrice** [comment]
Hey by the way @satyaog, the integration test is logging to this project, perhaps we could include a URL to it inside the guide or the wandb example? https://wandb.ai/lebrice/wandb-example

**Review by soline-b — APPROVED** [review]
